.github/workflows/test.yml

```diff
@@ -1,12 +1,40 @@
 name: Tests
 
-on: [push, pull_request]
+on:
+  push:
+    branches-ignore:
+      - tmp/*
+  pull_request:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
 
 # Thanks, https://swiftpackageindex.com/docs/builds for help coming up with *how* we build the artifacts. Swift build on Linux, XCode build on mac. Changing the XCode version to a specific version is how to set what version of Swift we want to test against on macOS. This is more real-world but also because I encountered issues with trying to install a Swift version in the macOS environment and try to run tests on it. 
 # See example commands by choosing a package: https://swiftpackageindex.com/nativedevbr/swift-log/builds and viewing the build information for each environment to see the commands run in that environment. 
 
 jobs:
+  # Running tests can be time consuming on the CI server. Only run them if there were files modified that we determine should run the tests. 
+  files-changed:
+    name: Detect what files changed
+    runs-on: ubuntu-latest
+    # Map a step output to a job output
+    outputs:
+      src: ${{ steps.changes.outputs.src }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: dorny/paths-filter@v2
+        id: changes
+        with:
+          # Are there files that you believe should trigger running tests? Add it to the list of them below in the src list. 
+          filters: |
+            src:
+              - 'Sources/**'
+              - 'Tests/**'
+
   older-xcode-test:
+    needs: files-changed
+    if: needs.files-changed.outputs.src == 'true' 
     strategy:
       matrix:
```

Comment:
Thank you for this, I didn't know and had to look up what it does 😄 
But from the example, I have seen, it looks like, we don't have to create two jobs, we can use 1 job and not duplicate the steps. You can also provide OS information in `matric`

```
matrix:
        os: [macos-10.14, macos-10.15]
        xcode-version: [11.3, 11.4]
```